src/package/packagePushUpgrade.ts

```diff
@@ -185,13 +201,14 @@ export class PackagePushUpgrade {
         .map(
           (job, index) =>
             // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
-            `Job ${index + 1} Error:${JSON.stringify(job?.sf__Error, null, 2)}`
+            `Job ${index + 1}: Error:${JSON.stringify(job?.sf__Error, null, 2)}`
         )
         .join('');
 
       await fs.writeFile(outputFile, errorContent, 'utf-8');
+      throw new Error(`Push upgrade failed, job errors have been written to file: ${outputFile}`);
     } catch (error) {
-      throw new Error('Error when saving job errors to file' + (error as Error).message);
+      throw new Error('Error when saving job errors to file. ' + (error as Error).message);
```

Comment:
Prefer using `SfError` from `@salesforce/core`.  It has a static `wrap()` method that is pretty useful especially to preserve original error stack and cause.

Also, the error thrown on line 209 will be caught here and the error will be misleading/confusing.